### ISSUE TITLE
Custom html files and compiling info

### ISSUE THREAD

**Issue by syrys** [ngbp/ngbp#318]
Hi,

What i need to do is to have multiple html files on the project. this is just a hypothetical scenario. But say if you wanted index2.html on the root. Since the build script only looks for *.tpl.html and adds them to templates, it simply ignores the index2.html.

So, if i were to have such file, how would i go about adding it to the build/compile script (note, i do not want it as a template, but simply copy it over). Im sure there must be some copy command that i can add somewhere, but would be great if someone can tell me where and how to do this (would be great if i can do this as a pattern too). cheers.

Secondly, lets say i want this said index2.html from above example to also be compiled the same way as index.html. For example:  i want to be able to put this: <%= version %> and have it populate the version number same as index.html on index2.html.

Thirdly, i can see that the build script actually compiles the *.tpl.html into a js file. However, i noticed that it leaves behind and does not populate things such as "<%= version %>" on these template files. Is there any way i can make it do this (without hacking around and saving these vars as JS objects on index.html and populating as a scope/model variable)?

Sorry, lots of questions.

Thanks in advance. :)

**Comment by PaulL1** [comment]
For the first part, look in Gruntfile.js, and search for index.html.  You'll find a copy block that does the copy.  Duplicating this should let you have another file.  This should also do the <%=version%> on it.

For string replacing in the templates, the actual concatenation is done by html2js, you can also see that in the gruntfile.  

This page: https://github.com/karlgoldstein/grunt-html2js suggests that html2js will take a process block, similar to that on the copy for index.html.  So probably it's possible to add a process there somehow that does string replacement.

**Comment by syrys** [comment]
Yup i was actually looking at the gruntfile where the index.html processing was. Its great, that should do the trick, but that means i have to do it for every file. Im sure i could make it slightly better with an array and a forloop, but is there an easier way to do it with a wildcard match? like /main.notemplate.html compiles into /main.html. 

is there a command i can use in the grunt file to get all the *.notemplate.html? I suppose i can actually use the "filterForJS" method and reconstruct that. TY for the help. I reckon this is probably functionality that others would benefit?

as for the html2js and injecting variables to templates, version number for example, cant seem to find anything obvious and easy on that link. will look more into it slowly (its not something i need, but i thought the question should be asked anyway).

**Comment by syrys** [comment]
on a side note, can someone suggest me a way to stop karma opening up on firefox as soon as i do "grunt watch". I have to restart grunt watch often due to some bug(?) when adding new directories, and this opening of karma is annoying me haha.

**Comment by PaulL1** [comment]
In the karma directory, in the karma-unit.tpl.js, right at the bottom is a list of browsers.  You can comment them all out, and then manually open a Firefox (or Chrome) and point it at localhost:9018 (or localhost:9019 if something is already on 9018).  This browser will then just stay there without opening and closing.

I do occasionally get a memory leak on Chrome, so if I just leave it open it eventually dies.  But you may not get that problem.

**Comment by syrys** [comment]
nah that crashes and does not continue to compile (looks like its waiting for the browser to open which there is none). I found an easier way just now. Commenting out the karma processes from the build.

Default Build function on gruntfile:
  grunt.registerTask( 'build', [
    'clean', 'html2js', 'jshint', 'coffeelint', 'coffee', 'less:build',
    'concat:build_css', 'copy:build_app_assets', 'copy:build_vendor_assets',
    'copy:build_appjs', 'copy:build_vendorjs', 'index:build', 'karmaconfig',
    'karma:continuous' 
  ]);

and just comment out the last 2 items: 'karmaconfig', 'karma:continuous'

or create your own duplicate process that does not use karma for when you really dont care about it.

:)

**Comment by PaulL1** [comment]
Content preview:  Yes, but if you open a browser on localhost:9018, then it
   doesn’t wait because there’s one there. :-). But if your problem is that you
    don’t want to wait for the unit tests, then commenting it out works too.
   But don’t get in the habit of not running your unit tests, that’ll end badly…..
    [...] 

 Content analysis details:   (-2.9 points, 5.0 required)

  pts rule name              description

---

 -1.0 ALL_TRUSTED            Passed through trusted hosts only via SMTP
 -1.9 BAYES_00               BODY: Bayes spam probability is 0 to 1%
                             [score: 0.0000]

Yes, but if you open a browser on localhost:9018, then it doesn’t wait because there’s one there.  :-).  But if your problem is that you don’t want to wait for the unit tests, then commenting it out works too.  But don’t get in the habit of not running your unit tests, that’ll end badly…..=

**Comment by syrys** [comment]
@PaulL1 : yeah fair enough. For some reason it seemt to open/close/open again every time i kill the grunt watch and start grunt watch again. I can easily do the tests as needed.

Here is my Gruntfile changes for anyone else that needs it (do it near the bottom where all the tasks are):

``` javascript
  grunt.registerTask( 'watch-no-karma', [
      'build-no-karma',
      'delta' ] );
  var buildVars = [
      'clean', 'html2js', 'jshint', 'coffeelint', 'coffee', 'less:build',
      'concat:build_css', 'copy:build_app_assets', 'copy:build_vendor_assets',
      'copy:build_appjs', 'copy:build_vendorjs', 'index:build'
  ];
  grunt.registerTask( 'build-no-karma', buildVars);
  grunt.registerTask( 'build', buildVars.concat(['karmaconfig', 'karma:continuous']));
```

and just run with: grunt watch-no-karma

**Comment by PaulL1** [comment]
It kills firefox and restarts it because karma is configured to automatically start the browser.  You fix that by commenting out the browser name near the bottom of karma-unit.tpl.js:

```
  browsers: [
  // "Chrome"
]
```

It will then expect a browser to be manually opened on port 9018, and the unit tests won’t run unless that browser is there.

**Comment by syrys** [comment]
yeah its a bit of a hassle for anyone who wants to work on the project even for something little requires to  open a browser window to start up karma. Your method works great, but still requires you to open up localhost:9018 on the browser otherwise the watch process does not continue.

For now, i will stick to my watch-no-karma task, its working great (and use the normal watch if i want to run karma) :)
